packages/world/src/systems/MineSystem.sol

```diff
@@ -176,6 +174,83 @@ contract MineSystem is System {
 }
 
 library MineLib {
+  function _mineBed(EntityId bedEntityId, Vec3 bedCoord) public {
+    EntityId sleepingPlayerId = BedPlayer._getPlayerEntityId(bedEntityId);
+    if (sleepingPlayerId.exists()) {
+      (EntityId forceFieldEntityId, ) = getForceField(bedCoord);
+      (, uint128 depletedTime) = updateMachineEnergy(forceFieldEntityId);
+      EnergyData memory playerData = updateSleepingPlayerEnergy(sleepingPlayerId, bedEntityId, depletedTime, bedCoord);
+      PlayerUtils.removePlayerFromBed(sleepingPlayerId, bedEntityId, forceFieldEntityId);
+
+      // Kill the player
+      // The player is not on the grid so no need to call killPlayer
+      Energy._setEnergy(sleepingPlayerId, 0);
+      addEnergyToLocalPool(bedCoord, playerData.energy);
+      notify(sleepingPlayerId, DeathNotifData({ deathCoord: bedCoord }));
+    }
+  }
+
+  function _requireMinesAllowed(
+    EntityId playerEntityId,
+    ObjectTypeId objectTypeId,
+    Vec3[] memory coords,
+    bytes calldata extraData
+  ) public {
+    for (uint256 i = 0; i < coords.length; i++) {
+      Vec3 coord = coords[i];
+      (EntityId forceFieldEntityId, EntityId fragmentEntityId) = getForceField(coord);
+      if (forceFieldEntityId.exists()) {
+        (EnergyData memory machineData, ) = updateMachineEnergy(forceFieldEntityId);
+        if (machineData.energy > 0) {
+          bytes memory onMineCall = abi.encodeCall(
+            IForceFieldFragmentProgram.onMine,
+            (forceFieldEntityId, playerEntityId, objectTypeId, coord, extraData)
+          );
+
+          // We know fragment is active because its forcefield exists, so we can use its program
+          ResourceId fragmentProgram = fragmentEntityId.getProgram();
+          if (fragmentProgram.unwrap() != 0) {
+            callProgramOrRevert(fragmentProgram, onMineCall);
+          } else {
+            callProgramOrRevert(forceFieldEntityId.getProgram(), onMineCall);
+          }
+        }
+      }
+    }
+  }
+}
+
+library MassReductionLib {
+  function _processMassReduction(
+    EntityId playerEntityId,
+    EntityId minedEntityId,
+    Vec3 playerCoord
+  ) public returns (uint128) {
+    // TODO: balancing, what should the proper mass and energy cost be?
+    uint128 massLeft = Mass._getMass(minedEntityId);
+    if (massLeft == 0) {
+      return massLeft;
+    }
+
+    (uint128 toolMassReduction, ) = useEquipped(playerEntityId, massLeft);
+
+    uint128 playerEnergyReduction = 0;
+
+    // if tool mass reduction is not enough, consume energy from player up to mine energy cost
+    if (toolMassReduction < massLeft) {
+      uint128 remaining = massLeft - toolMassReduction;
+      playerEnergyReduction = PLAYER_MINE_ENERGY_COST <= remaining ? PLAYER_MINE_ENERGY_COST : remaining;
+      decreasePlayerEnergy(playerEntityId, playerCoord, playerEnergyReduction);
+    }
+
+    uint128 totalEnergyReduction = playerEnergyReduction + toolMassReduction;
```

Comment:
why is the mass being added to the local pool? the mass is mined but its in the inventory and will appear back when its built. So should not be converted into energy, otherwise total energy in the world will increase

Reply:
Ah you mean when it is burned?

Reply:
Nvm I think I get it

Reply:
Nvm I think I get it

Reply:
wait I think I still don't get it lol the mass that gets removed from the tool has to go somewhere right? I understand not transferring the mass of the object being mined as it will eventually be placed / consumed, but the tool's mass has to go somewhere

Reply:
the tools mass will be returned to the world when it breaks --> ores used in crafting will respawn

Reply:
done